src/newTest/java/org/opensearch/test/framework/cluster/LocalOpenSearchCluster.java

```diff
@@ -0,0 +1,610 @@
+                    && nodeInfo.getInfo(HttpInfo.class).address() != null;
+
+            if (!clientNodes.isEmpty()) {
+                NodeInfo nodeInfo = clientNodes.get(0);
+                if (hasHttpAddressPredicate.test(nodeInfo)) {
+                    final TransportAddress his = nodeInfo.getInfo(HttpInfo.class).address().publishAddress();
+                    clusterInfo.httpPort = his.getPort();
+                    clusterInfo.httpHost = his.getAddress();
+                } else {
+                    throw new RuntimeException("no http host/port for client node");
+                }
+            } else if (!dataNodes.isEmpty()) {
+                for (NodeInfo nodeInfo : dataNodes) {
+                    if (hasHttpAddressPredicate.test(nodeInfo)) {
+                        final TransportAddress his = nodeInfo.getInfo(HttpInfo.class).address().publishAddress();
+                        clusterInfo.httpPort = his.getPort();
+                        clusterInfo.httpHost = his.getAddress();
+                        break;
+                    }
+                }
+            } else {
+                for (NodeInfo nodeInfo : nodes) {
+                    if (hasHttpAddressPredicate.test(nodeInfo)) {
+                        final TransportAddress his = nodeInfo.getInfo(HttpInfo.class).address().publishAddress();
+                        clusterInfo.httpPort = his.getPort();
+                        clusterInfo.httpHost = his.getAddress();
+                        break;
+                    }
+                }
+            }
+
+            for (NodeInfo nodeInfo : nodes) {
+                clusterInfo.httpAdresses.add(nodeInfo.getInfo(HttpInfo.class).address().publishAddress());
+            }
+        } catch (final OpenSearchTimeoutException e) {
+            throw new IOException("timeout, cluster does not respond to health request, cowardly refusing to continue with operations");
+        }
+        return clusterInfo;
+    }
+
+    @Override
+    public String toString() {
+        return "\nES Cluster " + clusterName + "\nmaster nodes: " + masterNodes + "\n  data nodes: " + dataNodes + "\nclient nodes: " + clientNodes
+                + "\n";
+    }
+
+    private static List<String> toHostList(Collection<Integer> ports) {
+        return ports.stream().map(port -> "127.0.0.1:" + port).collect(Collectors.toList());
+    }
+
+    private String createNextNodeName(NodeSettings nodeSettings) {
+        List<Node> nodes;
+        String nodeType;
+
+        if (nodeSettings.masterNode) {
+            nodes = this.masterNodes;
+            nodeType = "master";
+        } else if (nodeSettings.dataNode) {
+            nodes = this.dataNodes;
+            nodeType = "data";
+        } else {
+            nodes = this.clientNodes;
+            nodeType = "client";
+        }
+
+        return nodeType + "_" + nodes.size();
+    }
+
+    public class Node implements OpenSearchClientProvider {
```

Comment:
Move into its own file?

Reply:
Right now, the Node class depends on a couple of properties of the surrounding class, LocalOpenSearchCluster. Also, it will be never used in a context other than LocalOpenSearchCluster. For me, this sounds like a use-case for using an inner class.

What would be your criteria, when to use inner classes vs when not to?

Reply:
LocalOpenSearchCluster is nearly 500 lines, its hard to read and maintain code when it gets to this point.  Sometimes tight coupling makes these files easier to use/read, other times it makes it difficult.  I had a hard time parsing the relationship between the cluster and the node at a glance it might be worth separating them to make it cleaner.  What do you think?

Reply:
Well, this is a case where some complexity is not completely avoidable - we just have the chose where to put it. We have two options:

- Node as a separate class in the `test.framework` package. This lowers  the complexity of the LocalOpenSearchCluster class, but increases clutter of the whole package tree, as the Node class is getting an actual - well - node in the package tree.
- Node as an inner class in LocalOpenSearchCluster. You have higher complexity inside the class. On the other hand, the package tree is kept clean.


LocalOpenSearchCluster will be hopefully a class normal developers have to seldomly touch. I would hope, it needs to be touched in less than 5% of all cases. On the other hand, developers need to constantly navigate the package tree. That's why I would personally prefer the case which minimized the clutter of the package tree at the  cost of evelated complexity inside LocalOpenSearchCluster.

Possibly, we can also find other ways to reduce the complexity inside LocalOpenSearchCluster.

Reply:
This isn't a blocker, let's leave it as is for the moment and we can iterate in future pull requests